beacon-chain/blockchain/service.go

```diff
@@ -222,26 +226,228 @@ func (c *ChainService) blockProcessing(processedBlock chan<- *types.Block) {
+//
+//		# check state root
+//      if block.state_root == hash(state):
+//			return state
+//		else:
+//			return False  # or throw or whatever
+//
+func (c *ChainService) receiveBlock(block *types.Block) error {
+
+	blockhash, err := block.Hash()
 	if err != nil {
-		return fmt.Errorf("failed to get hash of block: %v", err)
+		return fmt.Errorf("could not hash incoming block: %v", err)
 	}
 
-	if c.enablePOWChain && !c.doesPoWBlockExist(block) {
-		return errors.New("proof-of-Work chain reference in block does not exist")
+	beaconState, err := c.beaconDB.GetState()
+	if err != nil {
+		return fmt.Errorf("failed to get beacon state: %v", err)
+	}
+
+	if block.SlotNumber() == 0 {
+		return errors.New("cannot process a genesis block: received block with slot 0")
+	}
+
+	// Save blocks with higher slot numbers in cache.
+	if !c.isBlockReadyForProcessing(block) && block.SlotNumber() > beaconState.Slot() {
+		c.unProcessedBlocks[block.SlotNumber()] = block
+		log.Debugf("block with hash %#x is not ready for processing", blockhash)
+		return nil
+	}
+
+	// Check for skipped slots and update the corresponding proposers
+	// randao layer.
+	for beaconState.Slot() < block.SlotNumber()-1 {
+		beaconState, err = c.executeStateTransition(beaconState, nil)
+		if err != nil {
+			return fmt.Errorf("unable to execute state transition %v", err)
+		}
+	}
+
+	beaconState, err = c.executeStateTransition(beaconState, block)
+	if err != nil {
+		return errors.New("unable to execute state transition")
+	}
+
+	// TODO(#1074): Verify block.state_root == hash_tree_root(state)
+	// if there exists a block for the slot being processed.
+
+	if err := c.beaconDB.SaveBlock(block); err != nil {
+		return fmt.Errorf("failed to save block: %v", err)
+	}
+	if err := c.beaconDB.SaveUnfinalizedBlockState(beaconState); err != nil {
+		return fmt.Errorf("error persisting unfinalized block's state: %v", err)
+	}
+
+	log.WithField("hash", fmt.Sprintf("%#x", blockhash)).Debug("Processed beacon block")
+
+	// We keep a map of unfinalized blocks in memory along with their state
+	// pair to apply the fork choice rule.
+	c.unfinalizedBlocks[blockhash] = beaconState
+
+	return nil
+}
+
+// Spec:
+//  We now define the state transition function. At a high level the state transition is made up of two parts:
+//  - The per-slot transitions, which happens every slot, and only affects a parts of the state.
+//  - The per-epoch transitions, which happens at every epoch boundary (i.e. state.slot % EPOCH_LENGTH == 0), and affects the entire state.
+//  The per-slot transitions generally focus on verifying aggregate signatures and saving temporary records relating to the per-slot
+//  activity in the BeaconState. The per-epoch transitions focus on the validator registry, including adjusting balances and activating
+//  and exiting validators, as well as processing crosslinks and managing block justification/finalization.
+func (c *ChainService) executeStateTransition(
```

Comment:
@nisdas can we move this to `core`?

Reply:
I guess we are not moving this to core?

Reply:
hmm , I already moved it to core, not sure why its showing it in service.go